### PR TITLE
feat: REST endpoint to delete dataset

### LINKED ISSUE #3155: [datasets] DELETE /datasets/:dataset_id

### PR DESCRIPTION
**Pull request by RogerHYang**
resolves #3155

### CODE REVIEW

File: src/phoenix/server/api/mutations/dataset_mutations.py [hunk 1]
```diff
@@ -274,21 +274,53 @@ async def delete_dataset(
...
+            try:
+                async with info.context.db() as session:
+                    await session.execute(
+                        delete(models.Project).where(
+                            models.Project.name.in_(set(project_names)),
+                        )
+                    )
+            except BaseException:
+                pass
+        if eval_trace_ids:
```
Comment (axiomofjoy) [inline src/phoenix/server/api/mutations/dataset_mutations.py:314]:
```suggestion
        if run_trace_ids:
```

Or `experiment_run_trace_ids`.

File: tests/server/api/routers/v1/test_datasets.py [hunk 2]
```diff
@@ -536,3 +537,12 @@ async def test_post_dataset_upload_pyarrow_create_then_append(test_client, sessi
     assert revisions[1].input == {"a": 11, "b": 22, "c": 33}
     assert revisions[1].output == {"b": 22, "c": 33, "d": 44}
     assert revisions[1].metadata_ == {"c": 33, "d": 44, "e": 55}
+
+
+async def test_delete_dataset(test_client, empty_dataset) -> None:
+    url = f"v1/datasets/{GlobalID(Dataset.__name__, str(1))}"
+    assert len((await test_client.get("/v1/datasets")).json()["data"]) > 0
+    (await test_client.delete(url)).raise_for_status()
```
Comment (axiomofjoy) [inline tests/server/api/routers/v1/test_datasets.py:545]: Suggestion: Test to ensure the deleted dataset id is returned.

File: src/phoenix/server/api/routers/v1/datasets.py [hunk 3]
```diff
@@ -163,6 +165,102 @@ async def list_datasets(request: Request) -> Response:
...
+        if (await session.scalar(stmt)) is None:
+            return Response(content="Dataset does not exist", status_code=HTTP_404_NOT_FOUND)
+    if not (project_names or eval_trace_ids):
+        return Response()
+    tasks = BackgroundTasks()
+    if project_names:
+        tasks.add_task(_delete_projects, request.app.state.db, frozenset(project_names))
+    if eval_trace_ids:
+        tasks.add_task(_delete_traces, request.app.state.db, frozenset(eval_trace_ids))
+    return Response(background=tasks)
```
Comment (axiomofjoy) [inline src/phoenix/server/api/routers/v1/datasets.py:233]: Would return at least the deleted dataset id as part of the response.

File: src/phoenix/server/api/routers/v1/datasets.py [hunk 4]
```diff
@@ -163,6 +165,102 @@ async def list_datasets(request: Request) -> Response:
...
+        return Response(
+            content="Missing Dataset ID",
+            status_code=HTTP_422_UNPROCESSABLE_ENTITY,
+        )
+    project_names_stmt = (
+        select(models.Experiment.project_name)
+        .where(models.Experiment.dataset_id == dataset_id)
+        .where(models.Experiment.project_name.isnot(None))
+    )
+    eval_trace_ids_stmt = (
```
Comment (axiomofjoy) [inline src/phoenix/server/api/routers/v1/datasets.py:211]:
```suggestion
    run_trace_ids_stmt = (
```

File: src/phoenix/server/api/routers/v1/datasets.py [hunk 5]
```diff
@@ -163,6 +165,102 @@ async def list_datasets(request: Request) -> Response:
...
+        tasks.add_task(_delete_traces, request.app.state.db, frozenset(eval_trace_ids))
+    return Response(background=tasks)
+
+
+async def _delete_projects(
+    db: Callable[[], AsyncContextManager[AsyncSession]],
+    project_names: FrozenSet[str],
+) -> None:
+    if not project_names:
+        return
```
Comment (axiomofjoy) [inline src/phoenix/server/api/routers/v1/datasets.py:241]:
```suggestion
```

Can nix given the other check.

File: src/phoenix/server/api/routers/v1/datasets.py [hunk 6]
```diff
@@ -163,6 +165,102 @@ async def list_datasets(request: Request) -> Response:
...
+    except BaseException:
+        pass
+
+
+async def _delete_traces(
+    db: Callable[[], AsyncContextManager[AsyncSession]],
+    trace_ids: FrozenSet[str],
+) -> None:
+    if not trace_ids:
+        return
```
Comment (axiomofjoy) [inline src/phoenix/server/api/routers/v1/datasets.py:255]:
```suggestion
```

**Review by axiomofjoy — APPROVED** [review]

File: src/phoenix/server/api/mutations/dataset_mutations.py (hunk 1 above)
Comment (RogerHYang) [reply src/phoenix/server/api/mutations/dataset_mutations.py:314]: the run trace ids are deleted by the projects

Comment (axiomofjoy) [reply src/phoenix/server/api/mutations/dataset_mutations.py:314]: Whoops, misread.

File: tests/server/api/routers/v1/test_datasets.py (hunk 2 above)
Comment (RogerHYang) [reply tests/server/api/routers/v1/test_datasets.py:545]: looks like it should be 204 https://www.vinaysahni.com/best-practices-for-a-pragmatic-restful-api#:~:text=Response%20to%20a%20successful%20request%20that%20won%27t%20be%20returning%20a%20body%20(like%20a%20DELETE%20request)

> 204 No Content - Response to a successful request that won't be returning a body (like a DELETE request)

File: src/phoenix/server/api/routers/v1/datasets.py (hunk 3 above)
Comment (RogerHYang) [reply src/phoenix/server/api/routers/v1/datasets.py]: i changed it to 204

**Pull request merged**